src/averageAges.js

```diff
@@ -1,64 +1,51 @@
- * function calculates average age only for men who died in this century
- *
- * To calculate century:
- * Divide year of person's death by 100: Math.ceil(person.died / 100)
- *
- * @param {object[]} people
- * @param {number} century - optional
- *
- * @return {number}
- */
+function calculateAverageAge(people) {
+  return people.reduce((acc, { died, born }) =>
+    acc + (died - born), 0) / people.length;
+}
+
+function isParent(people, person) {
+  return people.find(({ name }) => name === person.mother);
+}
+
+const calculateCentury = ({ died }) => Math.ceil(died / 100);
+
+const isMale = ({ sex }) => sex === 'm';
+const isFemale = ({ sex }) => sex === 'f';
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const menFilteredAges = people
+    .filter(person => {
+      return century
+        ? isMale(person)
+          && calculateCentury(person) === century
+        : isMale(person);
+    });
+
+  return calculateAverageAge(menFilteredAges);
 }
 
-/**
- * Implement calculateWomenAverageAge function
- *
- * Function returns average age of women in array. If `withChildren` is
- * specified then function calculates average age only for women with children
- *
- * Hint: To check if a woman has children you should find someone who mention
- * her as mother.
- *
- * @param {object[]} people
- * @param {boolean} withChildren - optional
- *
- * @return {number}
- */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const womenFilteredAges = people
+    .filter(person => withChildren
+      ? isFemale(person)
+      && people.find(({ mother }) => mother === person.name)
+      : isFemale(person)
+    );
+
+  return calculateAverageAge(womenFilteredAges);
 }
 
-/**
- * Implement calculateAverageAgeDiff function.
- *
- * The function returns an average age difference between a child and his or her
- * mother in the array. (A mother's age at child birth)
- *
- * If `onlyWithSon` is specified then function calculates age difference only
- * for sons and their mothers.
- *
- * @param {object[]} people
- * @param {boolean} onlyWithSon - optional
- *
- * @return {number}
- */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // 1. find a mother of each person (or only for men)
-  // 2. keep people who have mothers in the array
-  // 3. calculate the difference child.born - mother.born
-  // 4. return the average value
+  const filteredChildren = people
+    .filter(person => onlyWithSon
+      ? isParent(people, person)
+        && isMale(person)
+      : isParent(people, person))
+    .map((child) => child.born - isParent(people, child).born);
+
+  return filteredChildren.reduce((a, b) => a + b) / filteredChildren.length;
```

Comment:
Now everything is written correctly, but such code is difficult to read and maintain. Break it into smaller parts.
For example, you filter children by mother in any case, so you could do something like this:
```js
let filteredChildren = people.filter(person => findMother(people, person));
if (onlyWithSon) {
  filteredChildren = filteredChildren.filter(child => isMale(child));
}
etc.
```